tpu-client/src/nonblocking/tpu_client.rs

```diff
@@ -933,3 +896,38 @@ async fn maybe_fetch_cache_info(
         maybe_slot_leaders,
     }
 }
+
+/// Create LeaderTpuService
+/// Retries until successful or timeout is reached
+async fn create_leader_tpu_service<P, M, C>(
+    rpc_client: Arc<RpcClient>,
+    websocket_url: &str,
+    exit: Arc<AtomicBool>,
+) -> Result<LeaderTpuService>
+where
+    P: ConnectionPool<NewConnectionConfig = C>,
+    M: ConnectionManager<ConnectionPool = P, NewConnectionConfig = C>,
+    C: NewConnectionConfig,
+{
+    let start = Instant::now();
+    let tpu_leader_service_creation_timeout = Duration::from_secs(20);
+    loop {
+        if start.elapsed() > tpu_leader_service_creation_timeout {
+            return Err(TpuSenderError::Custom(format!(
+                "Failed to create LeaderTpuService connecting to: {}, timeout: {}s",
+                websocket_url,
+                tpu_leader_service_creation_timeout.as_secs()
+            )));
+        }
+
+        match LeaderTpuService::new(rpc_client.clone(), websocket_url, M::PROTOCOL, exit.clone())
+            .await
+        {
+            Ok(service) => return Ok(service),
+            Err(e) => match e {
+                TpuSenderError::RpcError(_) => sleep(Duration::from_secs(1)).await,
```

Comment:
Sorry to be even more annoying, but this might still hide errors.
Not every `RpcError` would be the error that we know we want to retry:

```rust
RpcError(
    Error {
        request: Some(GetSlotLeaders),
        kind: RpcError(RpcResponseError {
            code: -32602,
            message: "Invalid slot range: leader schedule for epoch 6 is unavailable",
            data: Empty
        })
    }
)
```

Also, I think, it would be better to move the timeout check into this branch.
And if the timeout has expired, we should print the last observed error.
If we do this, it is probably more forgiving if we hide some errors for a while, as long we still show the very last one.

This might still create a hard to debug issues, should the root cause be hidden and change during the timeout.
So, I think, it is still better if we would retry only on an error that is as specific to the one we know we need to retry on, as possible.